### ISSUE TITLE
 Integration of ioBroker.musiccast in MediaControl module

### ISSUE THREAD

**Issue by Zefau** [Zefau/ioBroker.jarvis#617]
_Originally posted by @zucki in https://github.com/Zefau/ioBroker.jarvis/issues/173#issuecomment-759701968_

> main:
> ![image](https://user-images.githubusercontent.com/4253035/104503364-3deee100-55e1-11eb-90a8-366d6000e711.png)

> netsub:
> ![image](https://user-images.githubusercontent.com/4253035/104503291-257ec680-55e1-11eb-8ba8-38be311091f7.png)

**Comment by Zefau** [comment]
```
{
	"_namespace": ".{PLAYER}",
	
	// information
	"artist": ".netusb.artist",
	"album": ".netusb.album",
	"track": ".netusb.track",
	"year": null,  // nicht verfügbar
	"cover": ".netusb.albumart_url",
	
	// actions
	"play": {
		"state": ".netusb.playback",
		"convert": val => val === 'play',
		"action": ".netusb.playPause"
	},
	"pause": {
		"action": ".netusb.playPause"
	},
	"resume": {
		"action": ".netsub.playPause"
	},
	"stop": {
		"action": ".netsub.stop"
	},
	"previous": {
		"action": ".netsub.prev"
	},
	"next": {
		"action": ".netsub.next"
	},
	
	// attributes
	"progress": {
		"state": null,
		"convert": null,
		"action": null,
		"setter": null
	},
	"duration": {
		"state": ".netusb.total_time"
	},
	
	// options
	"repeat": {
		"state": ".netsub.repeat_stat",
		"action": ".netsub.repeat"
	},
	"shuffle": {
		"state": ".netsub.shuffle_stat",
		"action": ".netsub.shuffle"
	},
	"volume": {
		"state": ".main.volume",
		"action": ".main.volume",
		"display": {
			"off": 0
		}
	}
}
```

**Comment by Zefau** [comment]
_Originally posted by @zucki in https://github.com/Zefau/ioBroker.jarvis/issues/173#issuecomment-759979456_

> Zum Thema Fortschrittsanzeige. Es gibt unter _netusb_ den Wert _play_time_. Allerdings gibt der zumindest bei Net-Radio und Spotify nicht den Trackfortschritt sondern die gesamte Zeit seit der letzten Pause an. Außerdem ist das Aktualisierungsinterval des musiccast Adapters nicht allzu hoch. Ich denke es macht wenig Sinn den Wert zu nutzen. Ich muss es mal noch mit lokal gespeicherter Musik probieren. _total_time_ bleibt bei mir immer 0. Also auch der Wert scheint nicht verwendbar zu sein.

> Ansonsten tut sich leider gar nichts.
> ![image](https://user-images.githubusercontent.com/4253035/104556596-2dba1e80-5640-11eb-8ad7-43f70a73ac05.png)

> Die Datenpunkte werden aber in dem Versuch befüllt.
> ![image](https://user-images.githubusercontent.com/4253035/104557087-ef712f00-5640-11eb-92d7-24ddaf252a04.png)

> Im Adapter kann es ja mehrere Geräte geben. Wie wird das denn im Widget dargestellt? Und beim anlegen des Mappings war mir ja nicht ganz klar was der Punkt "play" in den actions tut. Vielleicht ist da noch was falsch?
Es gibt übrigens den Datenpunkt .netusb.playback für jedes Gerät. Der wechselt zwischen Stop/Play/Pause je nach Status. Wird das Gerät ausgeschaltet wird auch Stop in den Datenpunkt geschrieben.

**Comment by Zefau** [comment]
@zucki verstehe ich das richtig, dass wenn du den PLAY-Button drückst sich ioBroker aktualisiert, aber nicht das MediaControl Modul? Das wäre äußerst merkwürdig.

Oder passiert generell sowohl bei den Buttons nichts als auch, wenn du über ioBroker startest?

**Comment by zucki** [comment]
@Zefau 
Da hab ich mich vielleicht etwas umständlich ausgedrückt. Ich wollte sagen dass die Datenpunkte im ioBroker aktualisiert werden wenn ich z.B. am Gerät Pause drücke und ich das Gerät auch über die Datenpunkte steuern kann.
Der Adapter funktioniert also.
Beim Jarvis Widget tut sich dagegen garnichts. Weder beim Druck auf einen Button noch wird es durch externe Trigger irgendwie aktualisiert.

**Comment by Zefau** [comment]
@zucki Ich sehe jetzt erst, dass bei dir die Player-Auswahl fehlt. Du musst einen Player auswählen, sonst geht es nicht. Da dieser aber nicht geladen wird, kannst du natürlich nichts auswählen.

Probier mal bitte `v2.2.0-beta.23`, ob es damit geht.

Geht es damit nicht, dann bräuchte ich die Objekt-Struktur des Adapters von dir, damit ich es richtig implementiert bekomme.

**Comment by zucki** [comment]
Wie komm ich denn an die beta.23? Bei der Installation über Github in ioBroker wird nur die beta.22 installiert.

Ansonsten scheint das Widget nun mit einem Ladesymbol hängen zu bleiben.
![image](https://user-images.githubusercontent.com/4253035/105381571-714bf400-5c0f-11eb-8939-8d93d8f4ea94.png)

**Comment by Zefau** [comment]
`v2.2.0-beta.23` ist nun oben.

**Comment by zucki** [comment]
@Zefau Danke, ich hab's gleich mal ausprobiert.
Leider bleibt das selbe Problem wie bei der Beta22: Das Widget zeigt einfach nur den drehenden Ladekreis und das war's.

**Comment by Zefau** [comment]
Was steht in der Browser Konsole?

**Comment by zucki** [comment]
Mhh, ich kenn mich in der Webentwicklung nicht so aus aber scheint nix besonderes drin zu sein.

![image](https://user-images.githubusercontent.com/4253035/105531187-36b68a00-5ce9-11eb-97c6-39f89a14feca.png)

**Comment by Zefau** [comment]
Kannst du mir mal den Objekt-Baum als json exportieren und an zefau@mailbox.org schicken?

**Comment by zucki** [comment]
Hab ich grad gemacht.
Danke für's weiterverfolgen.

**Comment by Zefau** [comment]
@zucki mit `v2.2.0-beta.24` solltest du jetzt ein Player-Auswahl haben und damit sollte das Widget nun hoffentlich funktionieren.

**Comment by zucki** [comment]
@Zefau Ja, Auswahl ist jetzt da und auch die Buttons, Titelanzeige und Albumcover funktionieren 👍. Ich werde die kommenden Tage mal alles durchtesten aber sieht gut aus.

**Comment by Zefau** [comment]
Wenn du es die Tage testest und es passt, dann schließe das Issue gerne.

**Comment by Zefau** [comment]
@zucki wie sieht's aus?

**Comment by zucki** [comment]
@Zefau 
Sorry, die letzten Wochen bin ich zu nicht's gekommen...
Drei Dinge sind mir aufgefallen die noch nicht sauber funktionieren:
Beim Abspielen einer Spotify Playlist lässt sich über die prev/next Buttons des Widgets nicht zwischen den Songs wechseln. Über den prev/next Datenpunkt im Adapter klappt es.
Auch der Repeat und Shuffle Button sind wirkungslos. Beim drüberfahren zeigen sie aber auch irgendwie _null_ an:
![image](https://user-images.githubusercontent.com/4253035/107874726-d10a8780-6ebb-11eb-9dc4-a01c41620bd4.png)

Außerdem ist das Auswahlfenster für das Gerät nicht groß genug. Schränkt die Funktion nicht ein, sieht aber nicht so schön aus.
![image](https://user-images.githubusercontent.com/4253035/107874698-91dc3680-6ebb-11eb-80de-7f9c9b8cc3f8.png)

Ansonsten läuft alles sauber. Danke auf jeden Fall für deine Arbeit.

**Comment by Zefau** [comment]
Also die prev/next bzw. repeat/shuffle Buttons gehen auf 
```
"previous": {
	"action": ".netsub.prev"
},
"next": {
	"action": ".netsub.next"
},

"repeat": {
	"state": ".netsub.repeat_stat",
	"action": ".netsub.repeat"
},
"shuffle": {
	"state": ".netsub.shuffle_stat",
	"action": ".netsub.shuffle"
},
```

Für mich schwierig, da ich es nicht testen kann.

**Comment by Zefau** [comment]
Welche Werte kann `repeat` bzw. `shuffle` annehmen?

**Comment by zucki** [comment]
Repeat und Shuffle haben die Rolle "Button". Weiß nicht wie es da mit den Werten aussieht.
Zusätzlich gibt es noch _repeat_stat_ und _shuffle_stat_ mit Rolle "text" die per "on/off" anzeigen ob Repeat oder Shuffle aktiviert sind.
Sieht man hier:
![image](https://user-images.githubusercontent.com/4253035/109394910-c4f5df80-7929-11eb-9c2d-4e539b5a795a.png)
